File: leetCode/python/p1_TwoSum/lc1.py
import time

class Solution:    
    """
    Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

    You may assume that each input would have exactly one solution, and you may not use the same element twice.

    You can return the answer in any order. 
    """ 
    def two_sum(self, nums: list, target: int) -> list: 
        sum = {}
        for i in range(len(nums)):
            difference = target - nums[i]    
            if difference in sum:
                return [sum.get(difference), i]
            else: 
                sum[nums[i]] = i

    def test_function(self, nums: list, target: int, sol: list, test_num: int): 
        answer = []
        start_time = time.time()
        answer = self.two_sum(nums, target) 
        end_time = time.time()
        if answer[0] == sol[0] and answer[1] == sol[1]:
            print('\nTEST' + str(test_num) + ': PASSED IN ' + '{:.2f}us'.format((end_time - start_time) * 1000000))
        else:
            print('\nTEST FAILED')        

s = Solution()

# Test 1

nums = [2,7,11,15]
target = 9
sol = [0, 1]

s.test_function(nums, target, sol, 1)

# Test 2

nums = [3, 2, 4]
target = 6
solution = [1, 2]

s.test_function(nums, target, solution, 2)

# Test 3

nums = [3, 3]
target = 6
solution = [0, 1]

s.test_function(nums, target, solution, 3)

File: leetCode/cpp/p1_TwoSum/lc1.cpp
/* Given an array of integers nums and an integer target, 
return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, 
and you may not use the same element twice.

You can return the answer in any order. */

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target);
    void printVector(vector<int>& vec);
};

vector<int> Solution::twoSum(vector<int>& nums, int target) {

    unordered_map<int,int> sum;

    for(int i = 0; i < nums.size(); i++) {
        int difference = target - nums[i];
        
        if(sum.find(difference) != sum.end())
            return {sum[difference], i};
        else
            sum[nums[i]] = i; 
    }

    return {};
}

void Solution::printVector(vector<int>& vec) {
   
    for(int i: vec) 
        cout << i << ' ';

    cout << '\n';
}

int main(void) {
   
    Solution s;
 
    /* Test 1: */ 
    cout << "Test 1:\n";

    vector<int> nums1 = {2, 7, 11, 15};
    int target1 = 9;

    cout << "Inputs 1: ";
    s.printVector(nums1);

    cout << "Target 1: " << target1 << '\n';

    vector<int> sol1;
    sol1 = s.twoSum(nums1, target1); 

    cout << "Solution 1: ";
    s.printVector(sol1);
    cout << '\n';

    /* Test 2: */
    cout << "Test 2:\n";

    vector<int> nums2 = {3, 2, 4};
    int target2 = 6;

    cout << "Inputs 2: ";
    s.printVector(nums2);

    cout << "Target 2: " << target2 << '\n';

    vector<int> sol2;
    sol2 = s.twoSum(nums2, target2); 

    cout << "Solution 2: ";
    s.printVector(sol2);
    cout << '\n';

    /* Test 3: */
    cout << "Test 3:\n";

    vector<int> nums3 = {3, 3};
    int target3 = 6;

    cout << "Inputs 3: ";
    s.printVector(nums3);

    cout << "Target 3: " << target3 << '\n';

    vector<int> sol3;
    sol3 = s.twoSum(nums3, target3); 

    cout << "Solution 3: ";
    s.printVector(sol3);
    cout << '\n';

    return EXIT_SUCCESS; 
}
